dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java

```diff
@@ -91,19 +87,25 @@ static ClassDumper create(List<ClassPathEntry> entries) throws IOException {
             .collect(toImmutableList());
     checkArgument(
         unreadableFiles.isEmpty(), "Some jar files are not readable: %s", unreadableFiles);
-
-    return new ClassDumper(entries, extensionClassLoader, mapJarToClassFileNames(entries));
+    
+    return new ClassDumper(entries, extensionClassLoader);
   }
 
   private ClassDumper(
       List<ClassPathEntry> inputClassPath,
-      ClassLoader extensionClassLoader,
-      ImmutableSetMultimap<ClassPathEntry, String> jarToClasses) {
+      ClassLoader extensionClassLoader) throws IOException {
     this.inputClassPath = ImmutableList.copyOf(inputClassPath);
     this.classRepository = createClassRepository(inputClassPath);
     this.extensionClassLoader = extensionClassLoader;
-    this.classPathEntryToClassFileNames = ImmutableSetMultimap.copyOf(jarToClasses);
-    this.classFileNameToClassPathEntry = ImmutableListMultimap.copyOf(jarToClasses.inverse());
+    
+    ImmutableListMultimap.Builder<String, ClassPathEntry> builder = ImmutableListMultimap.builder();
+    for (ClassPathEntry entry : inputClassPath) {
+      for (String className : entry.getClassNames()) {
+        builder.put(className, entry);
+      }
+    }
```

Comment:
Then the field isn't initialized in the constructor, in which case it shouldn't be a field. I do think we should get rid of this field completely, but one thing at a time.

Reply:
In the same way before this PR, the constructor can receive a map, which is built in the static factory method. It's a field because ClassDumper's non-static methods reference it.

Reply:
Yes, and that's a problem. This should not be a field. The class makes a hash of confusion between invocation data and object state. Invocation data like this map should be passed between methods, not stored in a field a pseudo-global state.

But we shouldn't fix everything in this PR. One field at a time.

Reply:
A ClassDumper works on a class path. I find storing the class path and symbol maps as fields concise than passing the variables among static methods. Especially, symbol maps are hidden from API (implementation details); the callers should not care about the symbol maps.

The map is not a pseudo-global state. For example, `ClassDumper.create([JarA, JarB]).findClassLocation(ClassX)` returns different result than `ClassDumper([JarC, JarD]).findClassLocation(ClassX)`.

Reply:
There are several questions in play here. One is what approach is easier to follow and maintain. I find the existing code difficult to follow and understand which is why I want to refactor it into more semantic classes with fewer (ideally zero) multimaps and other more complicated data structures.

The more abstract question is what *is* a ClassDumper? I think of it as a utility that can dump classes out of a classpath. Generally a utility doesn't operate on only one instance. E.g. the ExclusionFiles utility doesn't write a single exclusion file and then get thrown away. It's unsettling to see this as  something that is created for a single Classpath and then never used again. 

Indeed, that's is not an object oriented approach. A method that operates on only one object would typically be an instance method on the object's class. We don't have a Classpath class to put this in (though perhaps we should).Hence a utility method, but that should not be limited to a single object.

The current design does not take full advantage of object oriented programming. Rather than modelling things--classpaths, classpath entries, linkage errors, etc.--it's connecting them in an interlocking web of maps, lists, sets, lists of maps of sets, and so forth. This works at cross purposes to Java and makes the code hard to follow. I think using domain relevant classes like ClasspathEntry instead of generic data structures will make the underlying structure more apparent.

Reply:
1. I agree that domain relevant classes make the logic easy to follow. As you pointed out, our ClassPathEntry is better representation than `java.nio.file.Path`.

2. ClassDumper used to be a utility class with static methods before #231. The static methods took classLoader and BCEL's repository to operate. These parameters were always the same for the lifetime of LinkageChecker class:

(old code before #231)
```
    Class[] parameterTypes = ClassDumper.methodDescriptorToClass(methodSignature.getDescriptor(),
        classLoader);
```

```
                ClassDumper.listExternalMethodReferences(
                    className, jarFileToClasses, classLoaderFromJars, repository);
```

So it made sense to make ClassDumper non-static class. The situation has not changed; if we don't instantiate ClassDumper, there are the same set of parameters always passed to them. We may be able to abstract these parameters into a single class. Then we will want to add methods to the new class, resulting in creating another ClassDumper class with a different name.

Reply:
perhaps, but none of that changes that a single ClassDumper instance should be able to operate on multiple classpaths. Perhaps the extension class loader and some of the other parts can be configurable at an instance level, but the classpath shouldn't be. That's invocation level state.

Reply:
What do you think are invocation level state?

- `List<ClassPathEntry>`. Class path; list of JAR files.
- `FixedSizeClassPathRepository classRepository`, which loads BCEL's `JavaClass`es from `List<ClassPathEntry>`. It has cache for fast loading.
- `ClassLoader extensionClassLoader`, which loads classes from JDK, irrelevant to `List<ClassPathEntry>`
- `Multimap<String, ClassPathEntry> classFileNameToClassPathEntry`. To lookup JAR files that contains a class.

Reply:
The extensionClassLoader is part of the dumper since it's not derived from the classpath.

The other two are invocation state: they change when the classpath changes.

Reply:
If I interpret your explanation naively, having the two (classFileNameToClassPathEntry and classRepository) as invocation state would incur additional JAR file reading. For example, `classDumper.findClassLocation`, which currently leverages  classFileNameToClassPathEntry field, would require reading JAR files  every time LinkageChecker invokes the method. Am I interpreting your suggestion correctly?

Reply:
A lot of methods need to move around because they're not in the natural places now. That's probably one of them. I suspect we need a Classpath class rather than a List<ClasspathEntry> and findClassLocation belongs there.

The existing classes are an algorithm, not a model;  and that's why it's so hard to follow. I want this code to have a higher level abstraction where the algorithms can become internal, implementation details. Right now they're exposed on the surface of the API so everything is tightly coupled to everything else through shared data structures rather class types.

Reply:
Created an issue to followup https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/1346

Reply:
For this PR, would you move the IO part outside the constructor?

Doing the IO within the constructor (or outside) does not remove classFileNameToClassPathEntry field. Therefore it's not the reason to defy ["Avoid doing work in constructors"]("Avoid doing work in constructors").

Reply:
It's a purely formal move that adds an additional method argument so I'll do this with the expectation that the PR will then be merged without further changes.

Reply:
Thank you.